pika/adapters/select_connection.py

```diff
@@ -264,21 +264,38 @@ def get_next_deadline(self):
         timeout = min((timeout, SelectPoller.POLL_TIMEOUT))
         return timeout * SelectPoller.POLL_TIMEOUT_MULT
 
+    @staticmethod
+    def is_ready_to_fire(to_run):
+        """For patching by unittest"""
+        return True
+
     def process_timeouts(self):
         """Process the self._timeouts event stack"""
 
         now = time.time()
-        to_run = [timer for timer in self._timeouts.values()
-                  if timer['deadline'] <= now]
-
         # Run the timeouts in order of deadlines. Although this shouldn't
         # be strictly necessary it preserves old behaviour when timeouts
         # were only run periodically.
-        for t in sorted(to_run, key=itemgetter('deadline')):
-            t['callback']()
-            del self._timeouts[hash(frozenset(t.items()))]
+        to_run = sorted([(k,timer) for (k,timer) in self._timeouts.items()
+                         if timer['deadline'] <= now],
+                        key = lambda item : item[1]['deadline'])
+
+        if self.is_ready_to_fire(to_run):
+            # Force recalculation as the oldest timer is going to be removed.
             self._next_timeout = None
```

Comment:
Bug: `self._next_timeout = None` used to be inside the `for t in sorted ...` loop. It's a bug to have it here outside of the loop, because we don't know whether `to_run` list is empty or not at this point. Make it the first statement in your new `for k, timer ...` loop to make sure that it will be executed if `to_run` is not empty:

```
for k, timer in to_run:
    self._next_timeout = None

    if k not in self._timeouts:
        # Prior invocation must have deleted this timer
        continue
    ...
```